JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,536 @@
+*Checkout [`prop-types`](https://github.com/facebook/prop-types) lib on GitHub for general info.*
+
+In class components:
+- always use prop validation
+- use **static member** for declaring prop types and default props (if your transpiler doesn't support static members, set them after class declaration and then export the class; See: ["Exporting Components" section](#exporting-components))
+- use *defaultProps* if you need to set default values
+
+``` jsx
+export default class ComponentName extends React.Component {
+  static propTypes = {
+    id: PropTypes.string.isRequired,
+    name: PropTypes.string,
+  };
+
+  static defaultProps = {
+    name: 'John Doe',
+  };
+
+  // ...
+
+  render() {
+    const {name} = this.props;
+
+    return (
+      <div>{name}</div>
+    );
+  }
+}
+```
+
+In stateless components, use PropTypes for type checking but use native JS for setting default values.
+
+``` jsx
+function ComponentName({id, name = 'John Doe'}) {
+  // ...
+};
+
+ComponentName.propTypes = {
+  id: PropTypes.number.isRequired,
+  name: PropTypes.string
+}
+```
+
+### Using props
+
+If you are using more than one property in a method, preffer accessing them with **object destructuring**.
+
+``` jsx
+class ComponentName extends React.Component {
+  // ...
+  render() {
+    const {id, name, onClickHandler} = this.props;
+    // ...
+  }
+}
+```
+
+### Importing React
+``` jsx
+import * as React from 'react';
+
+class ComponentName extends React.Component {
+  // ...
+}
+```
+
+### Exporting Components
+Each React component file should have a single exported component.
+ 
+Exporting class components:
+``` jsx
+// preferred way of exporting
+export default class ComponentName extends React.Component {
+  // ...
+}
+
+// exporting when static members are not supported by the transpiler:
+class ComponentName extends React.Component {
+  // ...
+}
+
+ComponentName.propTypes = {
+  // ..
+};
+
+export default ComponentName;
+```
+
+Exporting stateless components:
+``` jsx
+// stateless components
+function PureComponent(/* props */) {
+  // ...
+}
+
+PureComponent.propTypes = {
+  // ...
+};
+
+export default PureComponent;
+
+// stateless components with no prop validation
+export default () => (/* ... */);
+```
+
+### Rendering Arrays of Data
+
+For iterating arrays:
+- use **map** method to render items
+- you should have a unique key for each element
+- **don't use array's index as a key**
+
+``` jsx
+render() {
+  // ...
+  return (
+    <ul>
+      {
+        booksArray.map((book) => {
+          <ListItem key={book.id} title={book.title} /> 
+        }); 
+      }
+    </ul>
+  );
+}
+```
+
+### Binding
+Don't bind functions inline, bind them in the class constructor.
+``` jsx
+// bad
+class ComponentName extends React.Component {
+  onClickHandler() {
+    // ...
+  }
+
+  render() {
+    return (
+      <ChildComponent onClick={this.onClickHandler.bind(this)} />
+    );
+  }
+}
+    
+// good
+class ComponentName extends React.Component {
```

Comment:
Maybe we should think about using something like https://github.com/andreypopp/autobind-decorator
@andreicek

Reply:
I really like this idea. How would do this with `@action.bound`? Would that still be the preferred way of binding actions or would we do this:

```javascript
@autobind
@action
doStuff() { /* code */ }
```

Here is a working example of this: https://codesandbox.io/s/v3jlrpy3w7

I actually think this is better since the methods usually start off not being actions.

Reply:
There is a known problem with `autobind-decorator` and TS. In the issue someone linked developit/decko module that has bind.

Here is an example of this running: https://codesandbox.io/s/ox0p2pym7z

Reply:
Looks good! Would avoid the `memoize` function trough, seems like it could cause unexpected bugs based on the way it caches the results.

Reply:
Agreed! I would then mention this lib and prefere it instead of `this.fn = this.fn.bind(this);`

@adrianbicanic what do you think?

Reply:
I can comment only regarding how the code looks. Def looks cleaner and more consistent than stuffing 6 methods bindings in the constructor and then binding some more action methods with decorators.

Reply:
OK - let it be so.

- use `developit/decko` lib
- `@bind` over constructor binding
- `@bind` over `@action.bound`
- don't use `memoize`